dbt/adapters/snowflake/relation_configs/catalog.py

```diff
@@ -0,0 +1,121 @@
+
+from dbt.adapters.snowflake.relation_configs.base import SnowflakeRelationConfigBase
+from dbt.adapters.snowflake.relation_configs.formats import TableFormat
+
+
+@dataclass(frozen=True, eq=True, unsafe_hash=True)
+class SnowflakeCatalogConfig(SnowflakeRelationConfigBase, RelationConfigValidationMixin):
+    """
+    This config follow the specs found here:
+    https://docs.snowflake.com/en/sql-reference/sql/create-iceberg-table
+    https://docs.snowflake.com/en/sql-reference/sql/create-dynamic-table#create-dynamic-iceberg-table
+
+    The following parameters are configurable by dbt:
+    - table_format: format for interfacing with the table, e.g. default, iceberg
+    - external_volume: name of the external volume in Snowflake
+    - base_location: the directory within the external volume that contains the data
+        *Note*: This directory can’t be changed after you create a table.
+
+    The following parameters are not currently configurable by dbt:
+    - name: snowflake
+    """
+
+    table_format: Optional[TableFormat] = TableFormat.default()
+    name: Optional[str] = "SNOWFLAKE"
+    external_volume: Optional[str] = None
+    base_location: Optional[str] = None
+
+    @property
+    def validation_rules(self) -> Set[RelationConfigValidationRule]:
+        return {
+            RelationConfigValidationRule(
+                (self.table_format == "default")
+                or (self.table_format == "iceberg" and self.base_location is not None),
+                DbtConfigError("Please provide a `base_location` when using iceberg"),
+            ),
+            RelationConfigValidationRule(
+                (self.table_format == "default")
+                or (self.table_format == "iceberg" and self.name == "SNOWFLAKE"),
+                DbtConfigError(
+                    "Only Snowflake catalogs are currently supported when using iceberg"
+                ),
+            ),
+        }
+
+    @classmethod
+    def from_dict(cls, config_dict: Dict[str, Any]) -> Self:
+        kwargs_dict = {
+            "name": config_dict.get("name"),
+            "external_volume": config_dict.get("external_volume"),
+            "base_location": config_dict.get("base_location"),
+        }
+        if table_format := config_dict.get("table_format"):
+            kwargs_dict.update({"table_format": TableFormat(table_format)})
+        return super().from_dict(kwargs_dict)
+
+    @classmethod
+    def parse_relation_config(cls, relation_config: RelationConfig) -> Dict[str, Any]:
+
+        if relation_config.config.extra.get("table_format") is None:
+            return {"table_format": "default"}
+
+        config_dict = {
+            "table_format": relation_config.config.extra.get("table_format"),
+            "name": "SNOWFLAKE",  # this is not currently configurable
+        }
+
+        if external_volume := relation_config.config.extra.get("external_volume"):
+            config_dict.update({"external_volume": external_volume})
```

Comment:
Why not just `config_dict["external_volume"]` since keys auto-add themselves and that's very pythonic.

Reply:
Oh do keys auto-add themselves on `extra`? I thought that was just a generic `dict`.

Reply:
Oh, I just realized what you meant.